libraries/core_libs/consensus/src/transaction/transaction_queue.cpp

```diff
@@ -40,31 +25,73 @@ std::shared_ptr<Transaction> TransactionQueue::get(const trx_hash_t &hash) const
-    return ret;
+SharedTransactions TransactionQueue::getOrderedTransactions(uint64_t count) const {
+  SharedTransactions ret;
+  ret.reserve(count);
+  uint64_t counter = 0;
+  std::unordered_map<addr_t, std::pair<std::map<val_t, std::shared_ptr<Transaction>>::const_iterator,
+                                       std::map<val_t, std::shared_ptr<Transaction>>::const_iterator>>
+      iterators;
+
+  // For accounts with multiple transactions we will iterate one level at a time
+  for (const auto &account : account_nonce_transactions_) {
+    iterators.insert({account.first, {account.second.begin(), account.second.end()}});
   }
 
-  std::vector<std::shared_ptr<Transaction>> ret;
-  ret.reserve(count);
-  size_t counter = 0;
-  for (const auto &trx : priority_queue_) {
-    if (counter == count) break;
-    ret.push_back(trx);
-    counter++;
+  // Retrieve gas_price ordered transactions for each unique account transaction with lowest nonce
+  while (true) {
+    std::multimap<val_t, std::shared_ptr<Transaction>> ordered_trxs;
+    for (auto it = iterators.begin(); it != iterators.end();) {
+      // Transactions ordered by gas price
+      if (it->second.first == it->second.second) {
+        // There are no more transactions for this account, account can be removed from iterators
+        it = iterators.erase(it);
+      } else {
+        ordered_trxs.insert({it->second.first->second->getGasPrice(), it->second.first->second});
+        // Increase iterator for an account
+        it->second.first++;
+        // Increase iterator to move to the next account
+        it++;
+      }
+    }
+    for (auto ordered_it = ordered_trxs.rbegin(); ordered_it != ordered_trxs.rend(); ordered_it++) {
+      ret.emplace_back(std::move(ordered_it->second));
+      counter++;
+      if (counter == count) break;
+    }
+    // Stop when we have the count or no more transactions in queue
+    if (counter == count || iterators.size() == 0) break;
   }
+
+  return ret;
+}
+
+SharedTransactions TransactionQueue::getAllTransactions() const {
+  SharedTransactions ret;
+  ret.reserve(queue_transactions_.size());
+  for (const auto &t : queue_transactions_) ret.push_back(t.second);
   return ret;
 }
 
 bool TransactionQueue::erase(const trx_hash_t &hash) {
   // Find the hash
-  const auto it = hash_queue_.find(hash);
-  if (it == hash_queue_.end()) return false;
+  const auto it = queue_transactions_.find(hash);
+  if (it == queue_transactions_.end()) return false;
+
+  const auto &account_it = account_nonce_transactions_.find(it->second->getSender());
+  if (account_it != account_nonce_transactions_.end()) {
+    const auto &nonce_it = account_it->second.find(it->second->getNonce());
+    if (nonce_it != account_it->second.end()) {
+      if (hash == nonce_it->second->getHash()) {
+        account_it->second.erase(nonce_it);
+        if (account_it->second.size() == 0) {
+          account_nonce_transactions_.erase(account_it);
+        }
+      }
+    }
+  }
```

Comment:
no need for all those nested ifs here, which are hard to read:

```
const auto it = queue_transactions_.find(hash);
if (it == queue_transactions_.end()) {
  return false;
}
queue_transactions_.erase(it);

const auto &account_it = account_nonce_transactions_.find(it->second->getSender());
if (account_it == account_nonce_transactions_.end()) {
  return
}

const auto &nonce_it = account_it->second.find(it->second->getNonce());
if (nonce_it == account_it->second.end()) {
  return
}
```

etc...

Reply:
Fixed